tool/src/js/runtime.mjs

```diff
@@ -67,68 +67,73 @@ export function enumDiscriminant(wasm, ptr) {
-    const ptr = wasm.diplomat_alloc(size, align);
-    encodeCallback(new Uint8Array(wasm.memory.buffer, ptr, size));
+  static slice = (wasm, list, rustType) => {
+    const elementSize = rustType == "u8" || rustType == "i8" || rustType == "bool" ? 1 :
+      rustType == "u16" || rustType == "i16" ? 2 :
+        rustType == "u64" || rustType == "i64" || rustType == "f64" ? 8 :
+          4;
 
-    this.ptr = ptr;
-    this.size = size;
-    this.free = () => {
-      const successfully_unregistered = DiplomatBuf_finalizer.unregister(this);
-      if (successfully_unregistered) {
-        wasm.diplomat_free(this.ptr, this.size, align);
-      } else {
-        console.error(`Failed to unregister DiplomatBuf at ${ptr}, this is a bug. Either it was never registered (leak), it was already unregistered (failed attempt to double free), or the unregister token was unrecognized (fallback to GC).`);
-      }
-    }
+    const byteLength = list.length * elementSize;
+    const ptr = wasm.diplomat_alloc(byteLength, elementSize);
+
+    // Create an array view of the buffer. This gives us the `set` method which correctly handles untyped values
+    const destination =
+      rustType == "u8" || rustType == "bool" ? new Uint8Array(wasm.memory.buffer, ptr, byteLength) :
+        rustType == "i8" ? new Int8Array(wasm.memory.buffer, ptr, byteLength) :
+          rustType == "u16" ? new Uint16Array(wasm.memory.buffer, ptr, byteLength) :
+            rustType == "i16" ? new Int16Array(wasm.memory.buffer, ptr, byteLength) :
+              rustType == "i32" || rustType == "isize" ? new Int32Array(wasm.memory.buffer, ptr, byteLength) :
+                rustType == "u64" ? new BigUint64Array(wasm.memory.buffer, ptr, byteLength) :
+                  rustType == "i64" ? new BigInt64Array(wasm.memory.buffer, ptr, byteLength) :
+                    rustType == "f32" ? new Float32Array(wasm.memory.buffer, ptr, byteLength) :
+                      rustType == "f64" ? new Float64Array(wasm.memory.buffer, ptr, byteLength) :
+                        new Uint32Array(wasm.memory.buffer, ptr, byteLength);
+    destination.set(list);
 
-    DiplomatBuf_finalizer.register(this, { wasm, ptr, size, align }, this);
+    return new DiplomatBuf(ptr, list.length, () => wasm.diplomat_free(ptr, byteLength, elementSize));
   }
 
-  leak = () => {
-    const successfully_unregistered = DiplomatBuf_finalizer.unregister(this);
-      if (successfully_unregistered) {
-        // leak
-      } else {
-        console.error(`Failed to unregister DiplomatBuf at ${this.ptr}, this is a bug. Either it was never registered (leak), it was already unregistered (failed attempt to double free), or the unregister token was unrecognized (fallback to GC).`);
-      }
+  constructor(ptr, size, free) {
+    this.ptr = ptr;
+    this.size = size;
+    this.free = free;
+    this.leak = () => { };
+    this.garbageCollect = () => DiplomatBufferFinalizer(this, this.free);
```

Comment:
Issue: What does `DiplomatBufferFinalizer(this, this.free)` do? Are you intending to do `DiplomatBufferFinalizer.register(this, this.free, this)` ?

Reply:
The last param is only required for unregistering

Reply:
I don't think this is necessary to be in the code. Shane only brought this up because I removed the param, but I don't think it's common in JS to document omitted optionals, they are treated more like overloads.